aten/src/ATen/native/Distance.cpp

```diff
@@ -159,13 +159,14 @@ Tensor _pdist_backward(const Tensor& grad, const Tensor& self, const double p, c
 }
 
 Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double eps) {
-  // Follow scipy impl to improve numerical precision
-  // Use x / sqrt(x * x) instead of x / (sqrt(x) * sqrt(x))
-  Tensor w12 = at::sum(x1 * x2, dim);
-  Tensor w1 = at::sum(x1 * x1, dim);
-  Tensor w2 = at::sum(x2 * x2, dim);
-  Tensor n12 = (w1 * w2).clamp_min_(eps * eps).sqrt_();
-  return w12.div_(n12);
+  auto x1_norm = at::norm(x1, 2, dim, true) + eps;
```

Comment:
Can we get a comment explaining the strategy here? It would also be good to compare it with the old implementation as well in the comment.

Reply:
Shall I touch upon how division by zero is handled?

Reply:
Ok, done!

Reply:
So you are avoiding overflow numerical problems of the original implementation, but you are exposing yourself to undeflow problems with yours (when normalizing, individual components may underflow and turn to 0). Do we have a feeling why one sort of tradeoffs is better than other?

Reply:
Underflow is not a problem, floats are the best when exponent is small.

Reply:
It is also one of the reasons why you normalized, say, imagines, before feeding them to a network...

Reply:
Sorry, underflow could be a problem, but I believe it is very unlikely. I have never seen a single example in my entire life where normalization would hurt. But, of course, I could be very wrong...

Reply:
Ah, sorry, I assume that this function is used with floating point types. Would that be a correct assumption?

Reply:
Yes, this function is used with floating point types. But you are right, usually underflow is less of a problem than overflow. It's possible to create an adversarial example though where instead of getting close to 1 cosine similarity you'd get a significantly smaller value.